text/056-debugger-gui.md

```diff
@@ -0,0 +1,103 @@
+an understandable way.  In addition, a visual debugger needs a panel to
+automatically display essential information such as the call stack and variable
+values.
+
+The PR for this CEP includes a mockup of the changes.  The changes add a new set of
+debugging tabs, identified by the name of associated toplevel script.  These
+tabs will be visible when one or more sessions have entered the debugger.
+Each debugger tab has two subtabs, one with the stack and variables, the other
+with the Messages panel.  The Messages panel is not shown in its usual
+location once the associated script has entered the debugger.
+
+Debugger tabs are created the first time the debugger stops for a session.
+
+The mockup is approximate; the look for the added panel will be like the rest of
+CoqIDE.
+
+TBD: The current layout suggests that the Errors and Jobs panels are one per
+script rather than what I would expect (one per CoqIDE process).  If so, they could
+be changed to be additional tabs next to the debugger tabs.
+
+While not part of this proposal, [#14040](https://github.com/coq/coq/issues/14040)
+would use less space for the Find/Replace panel.
+
+## Breakpoint consistency
+
+Maintaining breakpoint consistency after edits that move them
+is straightforward when the debugger stops only within the primary script.  
+CoqIDE can immediately update the location of breakpoints that appear
+after the last processed statement.  Edits changing the location of
+breakpoints before the last processed statement cause a roll back of
+the proof state, so these updates can also be applied immediately.
+(BTW, CoqIDE currently gets confused if the user edits before the last processed
+statement when coqidetop is busy.  This will be fixed by prohibiting edits
+to the script when coqidetop is busy.  See
+[#13967](https://github.com/coq/coq/issues/13967))
+
+In contrast, maintaining breakpoint consistency in secondary scripts is much harder.
+For example, if a user edits a secondary script in CoqIDE, giving new locations for
+breakpoints, those locations won't match those in the running session until the
+secondary script has been recompiled and the session has been restarted.
+While it may be feasible for CoqIDE to maintain and apply a mapping from original
+breakpoint locations to changed breakpoint locations, that seems complex and perhaps
+fragile.  Also, how do you highlight a stopping point when the associated statement
+has been deleted from the editor (i.e. from a "step", not a breakpoint)?
+
+Or consider the case in which two scripts both depend on `secondary.v`.
+After starting the first script, the user modifies and recompiles `secondary.v`
+and starts the second script.  To handle this case correctly, CoqIDE would need
+to maintain one mapping *per session*.
+
+On the other hand, we shouldn't go out of our way to support multiple
+simultaneous debug sessions--following a single debug session already requires
+very close attention by the user.
+

```

Comment:
Another detail: If two debug sessions stop in the same file, how do you highlight 2 stopping points?  The stopping point highlight would need to change based on the which debugger tab is active.